sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -1015,6 +1014,11 @@ case class Cast(
       } catch {
         case _: NumberFormatException => null
       }
+    case x: DayTimeIntervalType =>
+      buildCast[Long](_, dt => changePrecision(dayTimeIntervalToDecimal(dt, x.endField), target))
+    case x: YearMonthIntervalType =>
+      buildCast[Int](_, ym =>
+        changePrecision(Decimal(yearMonthIntervalToInt(ym, x.startField, x.endField)), target))
```

Comment:
shall we consider ansi mode here? what if the decimal precision is too small?

Reply:
yea let's add some negative tests

Reply:
I will take confusion on what is scale and what is precision to my grave....

We can round/truncate, but not modulo/wrap.
So I see:
CAST(INTERVAL '1.234' SECOND AS DECIMAL(1,0)) as perfectly fine.
But:
CAST(INTERVAL('123.4' SECOND AS DECIMAL(4, 2)) is an overflow.
On trunc vs round the semantic is given in DECIMAL to DECIMAL:
select cast(1.5 AS DECIMAL(1, 0));
2

PS: I loathe that cannot change decimal precision error. In most cases we should just call it an overflow.

Reply:
> PS: I loathe that cannot-change-decimal-precision error. In most cases we should just call it an overflow.

I plan to do that.